overwatch-derive/src/lib.rs

```diff
@@ -1,8 +1,70 @@
-mod utils;
-
+use proc_macro::TokenStream;
 use proc_macro_error2::{abort_call_site, proc_macro_error};
 use quote::{format_ident, quote};
-use syn::{punctuated::Punctuated, token::Comma, Data, DeriveInput, Field, Generics};
+use syn::{
+    parse, parse_macro_input, parse_str, punctuated::Punctuated, token::Comma, Data, DeriveInput,
+    Field, Fields, GenericArgument, Generics, ItemStruct, PathArguments, Type,
+};
+
+mod utils;
+
+#[expect(
+    clippy::missing_panics_doc,
+    reason = "We will add docs to this macro later on."
+)]
+#[proc_macro_attribute]
+pub fn derive_services(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(item as ItemStruct);
+    let struct_name = &input.ident;
+    let visibility = &input.vis;
+    let generics = &input.generics;
+
+    let Fields::Named(named_fields) = input.fields else {
+        panic!("`derive_services` macro only supports structs with named fields");
+    };
+    let fields = named_fields.named;
+
+    let modified_fields = fields.iter().map(|field| {
+        let field_name = &field.ident;
+        let field_type = &field.ty;
+
+        let Type::Path(path) = field_type else {
+            return quote! { #field };
+        };
+        let Some(segment) = path.path.segments.last() else {
+            return quote! { #field };
+        };
+        if segment.ident != "OpaqueServiceHandle" {
+            return quote! { #field };
+        }
+        let PathArguments::AngleBracketed(args) = &segment.arguments else {
+            return quote! { #field };
+        };
+        let new_args = {
+            let mut new_args = args.args.clone();
+            new_args.push(GenericArgument::Type(syn::parse_quote!(
+                AggregatedServiceId
+            )));
+            new_args
+        };
+
+        let new_type = quote! {
+            OpaqueServiceHandle<#new_args>
+        };
+
+        quote! { #field_name: #new_type }
+    });
+
+    // Generate the modified struct with #[derive(Services)]
+    let modified_struct = quote! {
+        #[derive(::overwatch::Services)]
+        #visibility struct #struct_name #generics {
+            #(#modified_fields),*
+        }
+    };
+
+    modified_struct.into()
+}
```

Comment:
Now that we use this. Im wondering is it wouldn't be simpler for the implementors to add just the type and forget about `OpaqueServiceHandle` at all:

```rust
#[derive_services]
pub stuct MyService {
    service_one: ServiceOneType,
    service_two: ServiceTwoType
}
```

Then we expand everything. I think this is cleaner. Wdyt?

Reply:
Oh yeah, that's a very good suggestion.

Reply:
Using the old `extract_type_from` did not work with your proposal, so I asked AI to help me support the new case without breaking the old one. I think it did a terrific job as everything seems to be generated correctly. Here's the commit: https://github.com/logos-co/Overwatch/pull/58/commits/2b18dbfe7cff51bf15f19600538d33dd43814e0b.

Reply:
Nice, this seems pretty good indeed.